modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteGlobalStatisticsManager.java

```diff
@@ -0,0 +1,985 @@
+                cfgMgr.checkLocalStatistics(cfg, topVer);
+                LocalStatisticsGatheringContext ctx = gatherer.gatheringInProgress(cfg.key());
+
+                if (ctx != null)
+                    // If there is no context = aggregation finished and data will be send at double check below
+                    ctx.futureAggregate().thenAccept(stat -> onLocalStatisticsAggregated(key, stat, topVer));
+            }
+
+            // Double check that we have no race with collection finishing.
+            objectStatistics = repo.getLocalStatistics(key, req.topVer());
+
+            if (checkStatisticsVersions(objectStatistics, req.versions())) {
+                StatisticsAddressedRequest removedReq = removeFromRequests(inLocalRequests, key, req.reqId());
+
+                if (removedReq != null)
+                    sendResponse(nodeId, removedReq.req().reqId(), key, StatisticsType.LOCAL, objectStatistics);
+                // else was already processed by on collect handler.
+            }
+        }
+    }
+
+    /**
+     * Test if statistics configuration is fit to all required versions.
+     * @param cfg Statistics configuration to check.
+     * @param versions Map of column name to required version.
+     * @return {@code true} if it is, {@code false} otherwise.
+     */
+    private boolean checkStatisticsCfg(StatisticsObjectConfiguration cfg, Map<String, Long> versions) {
+        if (cfg == null)
+            return false;
+
+        for (Map.Entry<String, Long> version : versions.entrySet()) {
+            StatisticsColumnConfiguration colCfg = cfg.columns().get(version.getKey());
+
+            if (colCfg == null || colCfg.version() < version.getValue())
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Test if specified statistics is fit to all required versions.
+     *
+     * @param stat Statistics to check.
+     * @param versions Map of column name to required version.
+     * @return {@code true} if it is, {@code false} otherwise.
+     */
+    private boolean checkStatisticsVersions(
+        ObjectStatisticsImpl stat,
+        Map<String, Long> versions
+    ) {
+        if (stat == null)
+            return false;
+
+        for (Map.Entry<String, Long> version : versions.entrySet()) {
+            ColumnStatistics colStat = stat.columnsStatistics().get(version.getKey());
+
+            if (colStat == null || colStat.version() < version.getValue())
+                return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Process incoming request for global statistics. Either response (if it exists), or collect and response
+     * (if current node is master node for the given key) or ignore (if current node is no more master node for
+     * the given key.
+     *
+     * @param nodeId Sender node id.
+     * @param req Request.
+     */
+    private void processGlobalRequest(UUID nodeId, StatisticsRequest req) throws IgniteCheckedException {
```

Comment:
multiple invocation of this method will trigger multiple collection of global statistics

Reply:
Add few checks to avoid it.